feat: enabling building cache

Farenheith opened the pull request:
Creating gradle building cache for the build pipeline

@Exopandora, you expressed the desire to have a branch-based cache. The problem with this approach is that, when the pipeline is triggered from a tag, it'll use the tag name instead, making a cache hit impossible for each release.


I've made the cache key to use the Minecraft version instead, I think it'll bring optimal hit for your case, and left a branch-based key as a fallback (you can see more details of how it works [here](https://github.com/marketplace/actions/cache).

I've also updated gradle version to match the documented version [here](https://docs.gradle.org/current/userguide/build_cache.html).

Farenheith commented:
I guess it worked. I forced an amend to make the build run again. The first one took 7 minutes, the second took 53s

Farenheith commented:
I think Now it's correct. But I want you to confirm something @Exopandora. Take a look at the artifact generated [here](https://github.com/Exopandora/ShoulderSurfing/actions/runs/14795406515). I think it has more than it needed. I want to update the pipeline to fix that, but I need to confirm with you first. Do you think the following paths to get the jars are enough?

```
            fabric/build/libs/*.jar
            forge/build/libs/*.jar
            neoforge/build/libs/*.jar
```

Exopandora commented:
Thanks for the PR. I will pushed some some changes to simplify it a bit, not sure if they work.

> Do you think the following paths to get the jars are enough?

Yes, only the three directories are needed:
```
fabric/build/libs/*.jar
forge/build/libs/*.jar
neoforge/build/libs/*.jar
```

Exopandora commented:
Caching still seems to work. It also makes sense to cache the
1. tmp and moddev folders generated by neoforge and neoform
2. and the loom-cache generated by fabric.
3. and the tmp fodler generated by forge.

Affected directories are:
```
/api/build/moddev
/api/tmp
/common/build/moddev
/common/tmp
/compat/build/moddev
/compat/tmp
/fabric/.gradle/loom-cache
/fabric/build/loom-cache
/forge/build/downloadMCMeta
/forge/build/tmp
/neoforge/build/moddev
/neoforge/tmp
```

I'll look into how to cache more folders, unless you beat me to it 😉.

Farenheith commented:
@Exopandora, looking further into the [documentation here](https://github.com/actions/setup-java/tree/v3/?tab=readme-ov-file#caching-packages-dependencies),  we can see that the default configuration caches **/*.gradle*, but it not seem to be version specific, so, maybe it's better to control cache entirely manually through the cache step

Exopandora commented:
> I see that you removed .build-cache from .gitignore. I recommend leaving it there because the local cache can influence the remote running. In the pipeline, I've used an action that caches that folder remotely right here

Gradle uses ~/.gradle for all its caching by default, which I also adjusted in the caching action:
```yaml
      - name: Cache Gradle Build
        id: cache-gradle-build
        uses: actions/cache@v4
        with:
          path: ~/.gradle
          key: ${{ runner.os }}-gradle-${{ github.ref_name }}-mc${{ steps.minecraft-version.outputs.minecraft_version }}
          restore-keys: |
            ${{ runner.os }}-gradle-${{ github.ref_name }}
```

But since its already done by the java step, it could perhaps be removed again?

> About the additional folders for caching, I wonder if they're necessary

They are used by the respective loaders as caches. For example, they store decompiled and patched minecraft jars, which are computationally expensive to generate.

Farenheith commented:
I understand, so, what I put in settings.gradle wasn't needed, right?

```
buildCache {
    local {
        directory = File(".build-cache")
        directory = rootDir.resolve(".build-cache")
    }
}
```

Exopandora commented:
> I understand, so, what I put in settings.gradle wasn't needed, right?
> 
> ```
> buildCache {
>     local {
>         directory = rootDir.resolve(".build-cache")
>     }
> }
> ```

Correct. Everything should be put in ~/.gradle by default and caching should be enabled by default.

Farenheith commented:
But take a look at the content of the folder. It's a bit different from .gradle one. I think it's some different cache

Exopandora commented:
> Se that the content was a bit different from .gradle folder:

Thanks for the pointer! I will take a closer look and let you know.

Exopandora commented:
According to the [documentation](https://docs.gradle.org/current/userguide/build_cache.html#sec:build_cache_enable):
> When the build cache is enabled, it will store build outputs in the Gradle User Home. 

Gradle user home is `~/.gradle` by default. Locally, I could find the files in `~/.gradle/build-cache-1`

Exopandora commented:
I prefer passing the `--build-cache` parameter explicitely in the pipeline.

Farenheith commented:
No problem! I've changed the PR and added the reconfigured the github actions cache to get all folders you listed, I'll force some commits now to see the cache working

Farenheith commented:
@Exopandora the problem here is that the cache is immutable, so, we can't keep a key fixed that will be updated in every run. We have some options to deal with it, though:

* Separate gradle cache and build cache
  For this option is better to move the build-cache folder to another one, and we can use a separate key for this cache that is based on the hash of all the source files of the project. For the base gradle cache (not build one), we can just use the default cache of setup-java described [here](https://github.com/actions/setup-java/tree/v3/?tab=readme-ov-file#caching-packages-dependencies)
* Force new cache using the trick described [here](https://github.com/actions/cache/blob/v4/tips-and-workarounds.md#update-a-cache)
  This approach is simpler to implement, but will consume cache quota for the repository. I don't think it's a problem because it'll adopt a LRU eviction when reaching limit, so all good.

I'll test the second option as the first one is similar for the first one I tried, with the exception that I didn't use a hash-based key. Also, the second one seems to have a better fit for your repository, as a hash based key, would never be used to accelerate builds between code source modifications

Exopandora commented:
Other articles only put the folders `~/.gradle/caches`and`~/.gradle/wrapper` into the cache, maybe the issue is related.

Farenheith commented:
Yeah, the current pipeline already uses Java setup cache with optimized key choosing, so removing it will not bring any improvement for the pipeline.  What we can do is to cache tmp folders, moddev and loom-cache based on a hash of gradle.properties and build.gradle files. What do you think?

Exopandora commented:
> based on a hash of gradle.properties and build.gradle files

I think it would be better to use gradle/libs.versions.toml to compute the cache key, because the build.gradle.kts files do not actually change and the gradle.properties only changes when the mod version changes.

Exopandora commented:
Doing clean builds ensures that no previous mod versions end up in the mod-jars artifact, and no outdated classes end up inside the jars.

Exopandora commented:
> See, in the first run, the default setup-java gradle cache restored lots of things and already made build take only 32s

This is about the time it takes to build locally.

Farenheith commented:
@Exopandora, I've updated the gradle cache step to make the hash over toml files, and here's what I got:

Time without custom cache (just set up Java one)
![image](https://github.com/user-attachments/assets/b715f851-0753-46bd-8dfe-c6fd01694d8c)
https://github.com/Exopandora/ShoulderSurfing/actions/runs/14811018418/job/41585681724

Time with both caches
![image](https://github.com/user-attachments/assets/f189405e-6c0c-4abf-a3a8-1aac56e53885)
https://github.com/Exopandora/ShoulderSurfing/actions/runs/14811026903/job/41585700917

Re-run to make sure:
![image](https://github.com/user-attachments/assets/61266961-953d-4b06-b58c-51b098333f8a)
https://github.com/Exopandora/ShoulderSurfing/actions/runs/14811037668/job/41585723011

Removing --build-cache to see if the time goes up
![image](https://github.com/user-attachments/assets/561cf736-2492-4b89-be19-a1826d8fe331)https://github.com/Exopandora/ShoulderSurfing/actions/runs/14811051111/job/41585749039

Re adding --build-cache:
![image](https://github.com/user-attachments/assets/27c02a3e-f4f6-40ac-a6e8-a28c441b9780)
https://github.com/Exopandora/ShoulderSurfing/actions/runs/14811077194/job/41585802733

See that, when having both caches being used, we apparently achieved 8 ~10 seconds less building time compared to just the default setup Java cache. I say apparently because the last step did go up again to 40s, but maybe it's just normal CPU oscillation, as those pipelines run in shared VMs.
I've also simplified the paths to obtain files to cache to make it less verbose, but I still get all the folders you requested.

I'm a bit skeptical as I don't have clear evidence of cache usage in the log, but the time is good and you no side effects will happen.

I think this is good to go, what do you think?

Exopandora commented:
@Farenheith I made the cacheable paths more explicit and changed the cache key generation back to using the minecraft version, because I think that way, the cache will provide the most value. You view the pipelines here:
1. I cleared all the caches and pushed the changes: https://github.com/Exopandora/ShoulderSurfing/actions/runs/14812408046/job/41588715262
2. Force pushed the same commit (with caches): https://github.com/Exopandora/ShoulderSurfing/actions/runs/14812481903/job/41588894033

If you think its gtg, I will merge it. Thanks for your commitment!

Farenheith commented:
@Exopandora if you use `${{ runner.os }}-gradle-cache-${{ steps.minecraft-version.outputs.minecraft_version }}` you'll have an updating problem with the cache key: it'll only be generated once per version and will never be updated again, as the cache is immutable. The key need to be based on the files it'll cache. You can have the minecraft_version in it, but the hash is essential

Exopandora commented:
Thanks for the detailed explanation. I reverted the change and updated the file list to include all files that influence the generated files.

Farenheith commented:
The artifacts generated now looks just enough, too.
![image](https://github.com/user-attachments/assets/fb1b9f02-1f3e-4a3f-9354-63c9b7e3c5e8)


I think this is good to go

Exopandora commented:
Thanks for verifying and contributing!
